support meta_tensor.to(device='cpu') under fake_mode

bdhirsh opened the pull request:
Fixing this is actually a bit annoying:

(1) FakeTensorMode sees a function where all of its inputs are real tensors, so it tries to run the real compute before converting the output to a FakeTensor

(2) we don't actually want this, because the "real compute" is support to error normally, when you do `meta_tensor.to(device='cpu')`. Instead, we want FakeTensor to actually skip constant prop and run the normal FakeTensor implementation, which will not error

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133044
* #146731
* __->__ #146729
* #146642

bdhirsh committed:
support meta_tensor.to(device='cpu') under fake_mode

[ghstack-poisoned]

zou3519 commented on test/test_fake_tensor.py:
```diff
@@ -2077,6 +2077,14 @@ def test_from_buffer(self):
+    def test_meta_tensor_to_fake_cpu(self):
+        x = torch.randn(4, 4, device='meta')
+        with FakeTensorMode(allow_non_fake_inputs=True):
+            breakpoint()
```
no breakpoint :P

bdhirsh replied:
whoops

bdhirsh committed:
Update on "support meta_tensor.to(device='cpu') under fake_mode"

Fixing this is actually a bit annoying:

(1) FakeTensorMode sees a function where all of its inputs are real tensors, so it tries to run the real compute before converting the output to a FakeTensor

(2) we don't actually want this, because the "real compute" is support to error normally, when you do `meta_tensor.to(device='cpu')`. Instead, we want FakeTensor to actually skip constant prop and run the normal FakeTensor implementation, which will not error




[ghstack-poisoned]

albanD reviewed (approved):
SGTM (except the breakpoint)

bdhirsh committed:
Update on "support meta_tensor.to(device='cpu') under fake_mode"

Fixing this is actually a bit annoying:

(1) FakeTensorMode sees a function where all of its inputs are real tensors, so it tries to run the real compute before converting the output to a FakeTensor

(2) we don't actually want this, because the "real compute" is support to error normally, when you do `meta_tensor.to(device='cpu')`. Instead, we want FakeTensor to actually skip constant prop and run the normal FakeTensor implementation, which will not error




[ghstack-poisoned]

bdhirsh committed:
Update on "support meta_tensor.to(device='cpu') under fake_mode"

Fixing this is actually a bit annoying:

(1) FakeTensorMode sees a function where all of its inputs are real tensors, so it tries to run the real compute before converting the output to a FakeTensor

(2) we don't actually want this, because the "real compute" is support to error normally, when you do `meta_tensor.to(device='cpu')`. Instead, we want FakeTensor to actually skip constant prop and run the normal FakeTensor implementation, which will not error




[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot --help
